include/xsparse/level_capabilities/co_iteration.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef XSPARSE_CO_ITERATION_HPP
+#define XSPARSE_CO_ITERATION_HPP
+
+#include <vector>
+#include <tuple>
+
+namespace xsparse::level_capabilities
+{
+    template <class IK, class PK, class Levels, class IterHelpers>
+    class Coiterate;
+
+    template <class IK, class PK, class... Levels, class... IterHelpers>
+    class Coiterate<IK, PK, std::tuple<Levels...>, std::tuple<IterHelpers...>>
+    {
+    public:
+        class coiteration_helper
+        {
+        private:
+            std::tuple<Levels...> levelsTuple;
+            std::tuple<IterHelpers...> iterHelpers;
+
+        public:
+            explicit inline coiteration_helper(std::tuple<IterHelpers...> iter_helpers,
+                                               Levels... levels)
+                : levelsTuple(std::make_tuple(levels...))
+                , iterHelpers(iter_helpers)
+            {
+            }
+
+            class iterator
+            {
+            private:
+                coiteration_helper const& m_coiterHelper;
+                std::tuple<IterHelpers::template iterator...> iterators;
+                IK min;
+
+            public:
+                using iterator_category = std::forward_iterator_tag;
+                using key_type = IK;
+                using pointer = std::pair<IK, std::vector<std::optional<PK>>>*;
+                using reference = std::tuple<IK, std::vector<std::optional<PK>>>;
+
+                explicit inline iterator(coiteration_helper const& coiterHelper,
+                                         std::tuple<IterHelpers::template iterator...> it)
+                    : m_coiterHelper(coiterHelper)
+                    , iterators(it)
+                {
+                    min = static_cast<IK>(std::get<0>(*std::get<0>(iterators)));
+                    std::apply([&](auto&&... args)
+                               { ((min = std::min(min, std::get<0>(*args))), ...); },
+                               iterators);
```

Comment:
Nice! I'll need to look into `std::apply`. I would change the argument name, it can get confusing with two `min`s.